### PR TITLE
[FLINK-8118] [table] Allow specifying reading offsets of KafkaTableSources

### PR DESCRIPTION
**Pull request by xccui**
## What is the purpose of the change

This PR enables the kafka table source builder to specify start reading offsets for the `KafkaTableSource`.

## Brief change log

  - Add offset specifying methods (reading from latest, earliest, group offsets and specified offsets) to the builder.
  - Configure the offset option in a new `getKafkaConsumer` and change the old `getKafkaConsumer` to `createKafkaConsumer`.
  - Add a related test and update the documents.

## Verifying this change

The change can be verified by `KafkaTableSourceTestBase. testKafkaTSSetConsumeOffsets`.

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (no)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (no)
  - The serializers: (no)
  - The runtime per-record code paths (performance sensitive): (no)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Yarn/Mesos, ZooKeeper: (no)
  - The S3 file system connector: (no)

## Documentation

  - Does this pull request introduce a new feature? (yes)
  - If yes, how is the feature documented? (docs)

### CODE REVIEW

File: flink-connectors/flink-connector-kafka-base/src/main/java/org/apache/flink/streaming/connectors/kafka/KafkaTableSource.java [hunk 1]
```diff
@@ -121,6 +130,37 @@ public String getProctimeAttribute() {
...
 
+	/**
+	 * Returns a version-specific Kafka consumer with the start position configured.
+	 *
+	 * @param topic                 Kafka topic to consume.
+	 * @param properties            Properties for the Kafka consumer.
+	 * @param deserializationSchema Deserialization schema to use for Kafka records.
+	 * @return The version-specific Kafka consumer
+	 */
+	public FlinkKafkaConsumerBase<Row> getKafkaConsumer(
```
Comment (fhueske) [inline flink-connectors/flink-connector-kafka-base/src/main/java/org/apache/flink/streaming/connectors/kafka/KafkaTableSource.java]: Should be `protected`

File: flink-connectors/flink-connector-kafka-base/src/main/java/org/apache/flink/streaming/connectors/kafka/KafkaTableSource.java [hunk 2]
```diff
@@ -163,14 +203,14 @@ protected void setRowtimeAttributeDescriptors(List<RowtimeAttributeDescriptor> r
 	//////// ABSTRACT METHODS FOR SUBCLASSES
 
 	/**
-	 * Returns the version-specific Kafka consumer.
+	 * Create a version-specific Kafka consumer.
```
Comment (fhueske) [inline flink-connectors/flink-connector-kafka-base/src/main/java/org/apache/flink/streaming/connectors/kafka/KafkaTableSource.java]: Create -> Creates

File: flink-connectors/flink-connector-kafka-base/src/main/java/org/apache/flink/streaming/connectors/kafka/KafkaTableSource.java [hunk 3]
```diff
@@ -68,6 +71,12 @@
 	/** Descriptor for a rowtime attribute. */
 	private List<RowtimeAttributeDescriptor> rowtimeAttributeDescriptors;
 
+	/** The startup mode for the contained consumer (default is {@link StartupMode#GROUP_OFFSETS}). */
+	protected StartupMode startupMode;
```
Comment (fhueske) [inline flink-connectors/flink-connector-kafka-base/src/main/java/org/apache/flink/streaming/connectors/kafka/KafkaTableSource.java]: should be `private`. We use `protected` setter methods to set optional parameters from the builder (see `proctimeAttribute`).

File: flink-connectors/flink-connector-kafka-base/src/main/java/org/apache/flink/streaming/connectors/kafka/KafkaTableSource.java [hunk 4]
```diff
@@ -309,6 +356,51 @@ public B withKafkaTimestampAsRowtimeAttribute(
 			return builder();
 		}
 
+		/**
+		 * Configures the TableSource to start reading from the earliest offset for all partitions.
+		 *
+		 * @see FlinkKafkaConsumerBase#setStartFromEarliest()
+		 */
+		public B startReadingFromEarliest() {
```
Comment (fhueske) [inline flink-connectors/flink-connector-kafka-base/src/main/java/org/apache/flink/streaming/connectors/kafka/KafkaTableSource.java]: shorten to `fromEarliestOffsets`?

File: flink-connectors/flink-connector-kafka-base/src/main/java/org/apache/flink/streaming/connectors/kafka/KafkaTableSource.java [hunk 5]
```diff
@@ -309,6 +356,51 @@ public B withKafkaTimestampAsRowtimeAttribute(
...
+			this.specificStartupOffsets = null;
+			return builder();
+		}
+
+		/**
+		 * Configures the TableSource to start reading from the latest offset for all partitions.
+		 *
+		 * @see FlinkKafkaConsumerBase#setStartFromLatest()
+		 */
+		public B startReadingFromLatest() {
```
Comment (fhueske) [inline flink-connectors/flink-connector-kafka-base/src/main/java/org/apache/flink/streaming/connectors/kafka/KafkaTableSource.java]: shorten to `fromLatestOffsets`?

File: flink-connectors/flink-connector-kafka-base/src/main/java/org/apache/flink/streaming/connectors/kafka/KafkaTableSource.java [hunk 6]
```diff
@@ -309,6 +356,51 @@ public B withKafkaTimestampAsRowtimeAttribute(
...
+			this.specificStartupOffsets = null;
+			return builder();
+		}
+
+		/**
+		 * Configures the TableSource to start reading from any committed group offsets found in Zookeeper / Kafka brokers.
+		 *
+		 * @see FlinkKafkaConsumerBase#setStartFromGroupOffsets()
+		 */
+		public B startReadingFromGroupOffsets() {
```
Comment (fhueske) [inline flink-connectors/flink-connector-kafka-base/src/main/java/org/apache/flink/streaming/connectors/kafka/KafkaTableSource.java]: shorten to `fromGroupOffsets`?

File: flink-connectors/flink-connector-kafka-base/src/main/java/org/apache/flink/streaming/connectors/kafka/KafkaTableSource.java [hunk 7]
```diff
@@ -309,6 +356,51 @@ public B withKafkaTimestampAsRowtimeAttribute(
...
+			return builder();
+		}
+
+		/**
+		 * Configures the TableSource to start reading partitions from specific offsets, set independently for each partition.
+		 *
+		 * @param specificStartupOffsets the specified offsets for partitions
+		 * @see FlinkKafkaConsumerBase#setStartFromSpecificOffsets(Map)
+		 */
+		public B startReadingFromSpecificOffsets(Map<KafkaTopicPartition, Long> specificStartupOffsets) {
```
Comment (fhueske) [inline flink-connectors/flink-connector-kafka-base/src/main/java/org/apache/flink/streaming/connectors/kafka/KafkaTableSource.java]: shorten to `fromSpecificOffsets`?

**Comment by xccui** [comment]
Hi @fhueske, thanks for the review! The PR has been updated according to your comments.

Thanks, Xingcan

File: docs/dev/table/sourceSinks.md [hunk 8]
```diff
@@ -264,6 +288,30 @@ val source: TableSource[_] = Kafka010AvroTableSource.builder()
 </div>
 </div>
 
+* **Specify the start reading position:** By default, the table source will start reading data from the committed group offsets in Zookeeper or Kafka brokers. You can specify other start positions via the builder's methods, which correspond to the configurations in section [Kafka Consumers Start Position Configuration](../connectors/kafka.html#kafka-consumers-start-position-configuration).
+
+<div class="codetabs" markdown="1">
+<div data-lang="java" markdown="1">
+{% highlight java %}
+TableSource source = Kafka010JsonTableSource.builder()
```
Comment (twalthr) [inline docs/dev/table/sourceSinks.md]: This should be Kafka010AvroTableSource.

Comment (xccui) [reply docs/dev/table/sourceSinks.md]: Oops, sorry about that and thanks for pointing it out.

**Pull request closed without merging**